app/main/views/service_updates.py

```diff
@@ -54,33 +56,24 @@ def service_status_update_audits(day=None, page=1):
 @login_required
 @role_required('admin', 'admin-ccs-category')
 def service_update_audits():
-    form = ServiceUpdateAuditEventsForm(request.args, csrf_enabled=False)
-
-    if not form.validate():
-        return render_template(
-            "service_update_audits.html",
-            today=datetime.utcnow().strftime(DATETIME_FORMAT),
-            acknowledged=form.default_acknowledged(),
-            audit_events=[],
-            form=form
-        ), 400
-
-    audit_events = data_api_client.find_audit_events(
+    audit_events = list(data_api_client.find_audit_events_iter(
```

Comment:
So.. this isn't really about table sizes. It's about the inappropriateness of the api endpoint that's being used here.

I agree that as written it has the potential for performance issues, or actually rather just general crapness. In the worst case scenario the oldest 1000 edits were made to the same service, and we spend, like, 10 pages retrieving event descriptions that we don't care about (we only care about the earliest for each service) just to end up listing a single service that has changes. And maybe being able to say "oh, and an unspecified number of other edited services...".

This is mostly because we're doing the grouping in the front-end and can't tell the api not to waste its time sending back events we don't care about. Being able to do the "grouping"  (actually it would behave more like `DISTINCT ON` but could be implemented using either `GROUP BY` or `DISTINCT ON`) on the server-side might initially be a little difficult to fit into the mental model of how the call works (I would propose a flag such as `earliest_for_each_object=true`), but would allow the superfluous events to be omitted. (whatever table the information ended up on, an endpoint behaviour like this would be required somewhere to avoid this problem)

When it comes to performance, table size concerns can be hugely mitigated by creating a partial index on `service_update` event types and that partial index can also serve the purpose of providing a sorted input for either `GROUP BY` or `DISTINCT ON` to operate on. If we're being really clever it might even be possible to perform the grouping in a subquery that uses an index-only-scan, meaning the heap tuples only ever need to be accessed for the earliest event of each object. Possibly only for the <=100 that are going to returned in this request. So these requests would by-and-large only have to read a small-table-sized index to perform the bulk of their work. And as many writes as the `AuditEvent` table gets, they're almost all `INSERT` and likely all end up grouped together in the same pages at the end of the table, leaving the visibility map comparatively clean, hopefully not hindering the ability to use an index only scan.

We actually already have an regular multi-column index on `AuditEvent` which would probably go some way towards making this query fast without any extra work. Even without an index only scan, an index can easily identify the `service_update` events.

If we want to look at reasons we get an api slowdown, table-size will be dwarfed by e.g. 
- use of `LIMIT .. OFFSET` as a paging mechanism, which only discards rows *after* it has pulled all the data for the possibly-hundreds-of-pages-long response out of the heap
- overzealous eager-joining, which is rife because we been defaulting to (I'm guilty of this too) eager joins on new relationships.
- the amount of time & memory we spend jsonifying & de-jsonifying data (json blobs in the tables need to be deserialized as they come out too). I'm not joking about this - json (de-)serialization is not particularly cheap and we would probably save significant amounts of both resources by just plugging in something such as https://pypi.python.org/pypi/ujson.

Of course, we're unable to talk about this truly scientifically and this text is full of "probably, hopefully..."s, because we don't have any proper profiling on a database with a live-like dataset on the same hardware. In an ideal world we'd be able to see exactly how much time a request spent performing each of these stages.

As to why *not* put the data on the `ArchivedService`?
- firstly it's more unnecessary denormalization of data we have elsewhere.
- it breaks a nice pattern we could set up for reviewing different types of events. It doesn't take the wildest mind to imagine a desire to monitor updates made to briefs, brief responses, supplier contact details... possibly even if we were to extend collaborative abilities to let suppliers see edits made to `DraftService`s during the submission phase, god knows...
- the interfaces are there. There currently isn't even a *basic* "list archived services" endpoint. That would be a lot of extra needless work to implement. And it's work that would need to be repeated any time we wanted to provide a similar change-monitoring ability for a new model. And inevitably it would be a different person implementing each one and we'd end up with something where none of those interfaces were *quite* alike for no good reason.